// solidarity/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.solidarity = {
    binaryExists: require('./extensions/functions/binaryExists'),
    getSolidaritySettings: require('./extensions/functions/getSolidaritySettings'),
    setSolidaritySettings: require('./extensions/functions/setSolidaritySettings'),
    checkRequirement: require('./extensions/functions/checkRequirement'),
    updateRequirement: require('./extensions/functions/updateRequirement'),
    skipRule: require('./extensions/functions/skipRule'),
    checkENV: require('./extensions/functions/checkENV'),
    checkCLI: require('./extensions/functions/checkCLI'),
    checkDir: require('./extensions/functions/checkDir'),
    checkFile: require('./extensions/functions/checkFile'),
    updateVersions: require('./extensions/functions/updateVersions'),
    removeNonVersionCharacters: require('./extensions/functions/removeNonVersionCharacters'),
    getVersion: require('./extensions/functions/getVersion'),
    getLineWithVersion: require('./extensions/functions/getLineWithVersion'),
    printResults: require('./extensions/functions/printResults'),
    reviewRule: require('./extensions/functions/reviewRule'),
    buildSpecificRequirement: require('./extensions/functions/buildSpecificRequirement'),
    appendSolidaritySettings: require('./extensions/functions/appendSolidaritySettings'),
    ruleHandlers: require('./extensions/functions/ruleHandlers'),
    createPlugin: require('./extensions/functions/createPlugin'),
};
var SolidarityOutputMode;
(function (SolidarityOutputMode) {
    SolidarityOutputMode[SolidarityOutputMode["MODERATE"] = 0] = "MODERATE";
    SolidarityOutputMode[SolidarityOutputMode["VERBOSE"] = 1] = "VERBOSE";
    SolidarityOutputMode[SolidarityOutputMode["SILENT"] = 2] = "SILENT";
})(SolidarityOutputMode = exports.SolidarityOutputMode || (exports.SolidarityOutputMode = {}));
