// gluegun/build/runtime/run.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var ramda_1 = require("ramda");
var toolbox_1 = require("../domain/toolbox");
var parameter_tools_1 = require("../toolbox/parameter-tools");
var runtime_find_command_1 = require("./runtime-find-command");
/**
 * Runs a command.
 *
 * @param rawCommand Command string or array of strings.
 * @param extraOptions Additional options use to execute a command.
 * @return The Toolbox object indicating what happened.
 */
function run(rawCommand, extraOptions) {
    if (extraOptions === void 0) { extraOptions = {}; }
    return __awaiter(this, void 0, void 0, function () {
        var toolbox, _a, command, array, _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    // use provided rawCommand or process arguments if none given
                    rawCommand = rawCommand || process.argv;
                    toolbox = new toolbox_1.Toolbox();
                    // attach the runtime
                    toolbox.runtime = this;
                    // parse the parameters initially
                    toolbox.parameters = parameter_tools_1.parseParams(rawCommand, extraOptions);
                    _a = runtime_find_command_1.findCommand(this, toolbox.parameters), command = _a.command, array = _a.array;
                    // jet if we have no command
                    if (ramda_1.isNil(command)) {
                        return [2 /*return*/, toolbox];
                    }
                    // rebuild the parameters, now that we know the plugin and command
                    toolbox.parameters = parameter_tools_1.createParams({
                        plugin: command.plugin.name,
                        command: command.name,
                        array: array,
                        options: toolbox.parameters.options,
                        raw: rawCommand,
                        argv: process.argv,
                    });
                    // set a few properties
                    toolbox.plugin = command.plugin || this.defaultPlugin;
                    toolbox.command = command;
                    toolbox.pluginName = toolbox.plugin.name;
                    toolbox.commandName = command.name;
                    // setup the config
                    toolbox.config = ramda_1.clone(this.config);
                    toolbox.config[toolbox.plugin.name] = ramda_1.merge(toolbox.plugin.defaults, (this.defaults && this.defaults[toolbox.plugin.name]) || {});
                    if (!toolbox.command.run) return [3 /*break*/, 2];
                    // allow extensions to attach themselves to the toolbox
                    this.extensions.forEach(function (extension) { return extension.setup(toolbox); });
                    // run the command
                    _b = toolbox;
                    return [4 /*yield*/, toolbox.command.run(toolbox)];
                case 1:
                    // run the command
                    _b.result = _c.sent();
                    _c.label = 2;
                case 2: return [2 /*return*/, toolbox];
            }
        });
    });
}
exports.run = run;
