// @storybook/react-native-server/dist/server/options.js
"use strict";

require("core-js/modules/es.object.assign");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _server = require("@storybook/core/server");

var _package = _interopRequireDefault(require("../../package.json"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function extendOptions(options, extendServer) {
  var manualId = options.manualId,
      secured = options.https,
      host = options.host,
      port = options.port;
  var storybookOptions = {
    manualId: manualId,
    secured: secured,
    host: host,
    port: port
  };
  return Object.assign({}, options, {
    framework: 'react-native',
    extendServer: extendServer,
    packageJson: _package["default"],
    mode: 'dev',
    ignorePreview: true,
    corePresets: [{
      name: _server.managerPreset,
      options: {
        managerEntry: require.resolve('../client/manager')
      }
    }, {
      name: require.resolve('./rn-options-preset.js'),
      options: {
        storybookOptions: storybookOptions
      }
    }]
  });
}

var _default = extendOptions;
exports["default"] = _default;
